File: switcher/internal/protocol/gateway.go
package protocol

import (
	"strings"

	"github.com/clovapi/switcher/internal/apistyle"
)

// UpstreamHints is the upstream-side metadata used by Electron enrichIrRequest.
type UpstreamHints struct {
	Model  string // effective wire model; overrides IR when non-empty.
	Source string // e.g. subscription:codex, subscription:claude-code, local:ollama, model:name
}

// GatewayEnrich mutates IR like electron/protocol/gateway.enrichIrRequest.
func GatewayEnrich(r *Request, h UpstreamHints) {
	model := strings.TrimSpace(h.Model)
	if model != "" {
		r.Model = model
	}
	switch strings.TrimSpace(h.Source) {
	case "subscription:codex":
		meta := ensureMeta(r)
		meta.CodexSubscription = true
	case "subscription:claude-code":
		meta := ensureMeta(r)
		meta.SubscriptionClaudeOAuth = true
	}
}

// ResolveUpstreamPath returns the ingress-stripped suffix for the upstream HTTP path.
func ResolveUpstreamPath(egress apistyle.Style, _ Request, upstreamSource string) string {
	return DefaultUpstreamPathSuffix(egress, upstreamSource)
}

File: electron/protocol/gateway.js
const { createIrRequest } = require("./ir");
const { defaultUpstreamPathSuffix } = require("./paths");

/**
 * @param {import('./ir').IrRequest} ir
 * @param {{ model?: string, api_style?: string }} upstream
 */
function enrichIrRequest(ir, upstream) {
  const model = String(upstream?.model || "").trim();
  // 代理路径已选定模型；覆盖客户端 body 里的 model（避免 Claude Code 填 Anthropic 名导致上游错模）。
  if (model) ir.model = model;
  if (ir.stream == null) ir.stream = true;
  if (upstream?.source === "subscription:codex") {
    ir.metadata = { ...(ir.metadata || {}), codexSubscription: true };
  }
  if (upstream?.source === "subscription:claude-code") {
    ir.metadata = { ...(ir.metadata || {}), subscriptionClaudeOAuth: true };
  }
  return ir;
}

function resolveUpstreamPath(egressStyle, ir, upstream) {
  return defaultUpstreamPathSuffix(egressStyle, upstream);
}

module.exports = {
  enrichIrRequest,
  resolveUpstreamPath,
};
